File: frontend/lib/models/notification_model.dart
class NotificationModel {
  final String id;
  final String type;
  final String title;
  final String body;
  final NotificationData? data;
  final DateTime createdAt;

  NotificationModel({
    required this.id,
    required this.type,
    required this.title,
    required this.body,
    this.data,
    required this.createdAt,
  });

  factory NotificationModel.fromJson(Map<String, dynamic> json) {
    return NotificationModel(
      id: json['_id'],
      type: json['type'] ?? 'General',
      title: json['title'] ?? '',
      body: json['body'] ?? '',
      data:
          json['data'] != null ? NotificationData.fromJson(json['data']) : null,
      createdAt: DateTime.parse(json['createdAt']),
    );
  }
}

class NotificationData {
  final String? employeeName;
  final String? employeePhoto;
  final String? image; // Selfie
  final String? address;
  final double? lat;
  final double? lng;
  final DateTime? eventTime;

  NotificationData({
    this.employeeName,
    this.employeePhoto,
    this.image,
    this.address,
    this.lat,
    this.lng,
    this.eventTime,
  });

  factory NotificationData.fromJson(Map<String, dynamic> json) {
    return NotificationData(
      employeeName: json['employeeName'],
      employeePhoto: json['employeePhoto'],
      image: json['image'],
      address: json['address'],
      lat: json['lat'] != null ? double.tryParse(json['lat'].toString()) : null,
      lng: json['lng'] != null ? double.tryParse(json['lng'].toString()) : null,
      eventTime:
          json['eventTime'] != null ? DateTime.parse(json['eventTime']) : null,
    );
  }
}

File: backend/models/notificationModel.js
import mongoose from "mongoose";

const notificationSchema = new mongoose.Schema({
  companyId: { type: mongoose.Schema.Types.ObjectId, ref: 'Company', required: true },
  // Removed 'recipientId' requirement because we are saving it generally for the company admins
  recipientId: { type: mongoose.Schema.Types.ObjectId, ref: 'User' }, 
  type: { 
    type: String, 
    enum: ['Attendance', 'Leave Request', 'Notes', 'Live Track', 'Payments', 'Announcement'], 
    required: true 
  },
  title: { type: String, required: true },
  body: { type: String, required: true },
  data: {
    employeeId: { type: mongoose.Schema.Types.ObjectId, ref: 'Employee' },
    employeeName: String,
    employeePhoto: String,
    image: String, // Selfie URL
    address: String,
    lat: Number,
    lng: Number,
    eventTime: Date,
    status: String // 'In', 'Out'
  },
  isRead: { type: Boolean, default: false },
}, { timestamps: true });

const Notification = mongoose.model('Notification', notificationSchema);
export default Notification;